main/sdr.go

```diff
@@ -319,21 +314,164 @@ func sdrKill() {
+	//         UAT    ES    id 1 -> ES  second pass loop 1, id 0 -> Unconfigured
+	//         ES     ES    id 0 -> ES  first  pass loop 1, id 1 -> unconfigured
+	//         UAT    UAT   id 0 -> Unconfigured          , id 1 -> Unconfigured
+
+	// es_enabled == false and uat_enabled == true
+	//
+	// ID      0      1
+	// -------------------------------------------------------------------------
+	//         A      A     id 0 -> UAT first  pass loop 2, id 1 -> Unconfigured
+	//         A      ES	id 0 -> UAT first  pass loop 2, id 1 -> Unconfigured
+	//         ES     A     id 1 -> UAT second pass loop 2, id 0 -> Unconfigured
+	//         A      UAT   id 1 -> UAT second pass loop 1, id 0 -> Unconfigured
+	//         UAT    A     id 0 -> UAT first  pass loop 1, id 1 -> Unconfigured
+	//         ES     UAT   id 1 -> UAT second pass loop 1, id 0 -> Unconfigured
+	//         UAT    ES    id 0 -> UAT first  pass loop 1, id 1 -> Unconfigured
+	//         ES     ES    id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         UAT    UAT   id 0 -> UAT first  pass loop 1, id 1 -> unconfigured
+
+	// es_enabled == false and uat_enabled == false
+	//
+	// ID      0      1
+	// -------------------------------------------------------------------------
+	//         A      A     id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         A      ES	id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         ES     A     id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         A      UAT   id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         UAT    A     id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         ES     UAT   id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         UAT    ES    id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         ES     ES    id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         UAT    UAT   id 0 -> Unconfigured          , id 1 -> Unconfigured
+
+	// once the tagged dongles have been assigned explicitly range over
+	// the remaining IDs and assign then to any anonymous dongles
+	unusedIDs := make([]int, 0, count)
+
+	// loop 1: assign tagged dongles
+	for i := 0; i < count; i++ {
+		_, _, s, err := rtl.GetDeviceUsbStrings(i)
+		if err == nil {
+			if uat_enabled && UATDev == nil && rUAT.hasID(s) {
+				createUATDev(i, s, true)
+			} else if es_enabled && ESDev == nil && rES.hasID(s) {
+				createESDev(i, s, true)
+			} else {
+				unusedIDs = append(unusedIDs, i)
+			}
+		} else {
+			log.Printf("rtl.GetDeviceUsbStrings id %d: %s\n", i, err)
+		}
 	}
 
+	// loop 2; assign anonymous dongles
+	for _, v := range unusedIDs {
+		_, _, s, err := rtl.GetDeviceUsbStrings(v)
+		if err == nil {
+			if uat_enabled && UATDev == nil {
+				createUATDev(v, s, false)
+			} else if es_enabled && ESDev == nil {
+				createESDev(v, s, false)
+			}
+		} else {
+			log.Printf("rtl.GetDeviceUsbStrings id %d: %s\n", v, err)
+		}
+	}
+}
```

Comment:
On Mon, Feb 22, 2016 at 12:31 PM, yihchun <notifications@github.com> wrote:

> In main/sdr.go
> <https://github.com/cyoung/stratux/pull/279#discussion_r53668876>:
>
> >  	}
> >
> > +	// loop 2; assign anonymous dongles
> > +	for _, v := range unusedIDs {
> > +		_, _, s, err := rtl.GetDeviceUsbStrings(v)
> > +		if err == nil {
> > +			if uat_enabled && UATDev == nil {
> > +				createUATDev(v, s, false)
> > +			} else if es_enabled && ESDev == nil {
> > +				createESDev(v, s, false)
> > +			}
> > +		} else {
> > +			log.Printf("rtl.GetDeviceUsbStrings id %d: %s\n", v, err)
> > +		}
> > +	}
> > +}
>
> one caveat is that we only support up to two dongles
>
> OK, I forgot about that force count = 2. I thought you were going to fix
> #239 <https://github.com/cyoung/stratux/issues/239> also. At this point I
> don't see any significant reason to do that force.
>
> Once loop 1 finishes the only remaining *enabled* dongles *must* be
> anonymous
>
> I'm not sure I see it that way. If both id 0 and id 1 have rES.hasID(s),
> then in round 1, id 0 will be wired to ES, then the next iteration id 1
> will not satisfy ESDev == nil so it will fall through to the else, thus
> being added to the unusedIDs array. At this point you'll send that one to
> UAT, which is correct behavior but id 1 is certainly not anonymous.
>

Lol. Okay, it only took you what 4 or 5 times to explain the dual
assignment scenario...thanks! I'll add the additional checks. :)